File: Cpp/315.cpp
class Solution {
public:
    vector<int> countSmaller(vector<int>& nums) {
        vector<int> res(nums.size(), 0);
        vector<int> fake_bst;
        for (int i = nums.size() - 1; i >= 0; i--) {
            int left = 0, right = fake_bst.size();
            while (left < right) {
                int mid = (left + right) / 2;
                if (nums[i] > fake_bst[mid])
                    left = mid + 1;
                else
                    right = mid;
            }
            // not a traditional binary search tree
            // https://en.wikipedia.org/wiki/Binary_search_algorithm
            res[i] = right;
            fake_bst.insert(fake_bst.begin() + right, nums[i]);
        }
        return res;
    }
};

class Solution {
public:
    void add(vector<int>& bit, int i) {
        while (i < bit.size()) {
            bit[i]++;
            i += i & -i;
        }
    }
    int sum(vector<int>& bit, int i) {
        int ans = 0;
        while (i > 0) {
            ans += bit[i];
            i -= i & -i;
        }
        return ans;
    }
    vector<int> countSmaller(vector<int>& nums) {
        vector<int> arr(nums), bit(nums.size() + 1, 0);
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size(); ++i)
            arr[i] = lower_bound(nums.begin(), nums.end(), arr[i])
                     - nums.begin() + 1;
        for (int i = nums.size() - 1; i >= 0; --i) { // cout<<arr[i]<<endl;
            add(bit, arr[i]);
            arr[i] = sum(bit, arr[i] - 1);
        }
        return arr;
    }
};

File: Python/315.py
class Solution(object):
    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        res = [0]*n
        bst = []
        for i in range(n-1, -1, -1):
            left, right = 0, len(bst)
            while left<right:
                mid = (left+right)/2
                if bst[mid] < nums[i]:
                    left = mid+1
                else:
                    right = mid
            res[i] = right
            bst.insert(right, nums[i])
        return res

# Merge sort
class Solution(object):
    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = [0]*len(nums)
        s = []
        for i, x in enumerate(nums):
            s.append((x, i))
        self.mergeSort(s, 0, len(s)-1, res)
        return res

    def mergeSort(self, s, l, r, res):
        if l>=r:
            return
        mid = (l+r)/2
        self.mergeSort(s, l, mid, res)
        self.mergeSort(s, mid+1, r, res)
        self.merge(s, l, mid, r, res)

    def merge(self, s, l, mid, r, res):
        i, j = l, mid+1
        num_of_right_lt_left = 0
        t = []
        while i<=mid and j<=r:
            if s[i][0] <= s[j][0]:
                t.append(s[i])
                res[s[i][1]] += num_of_right_lt_left
                i+=1
            else:
                t.append(s[j])
                num_of_right_lt_left +=1
                j+=1
        while i<=mid:
            t.append(s[i])
            res[s[i][1]] += num_of_right_lt_left
            i+=1
        while j<=r:
            t.append(s[j])
            j+=1
        for i, x in enumerate(t):
            s[l+i] = x


